test/object-store/sync/flx_role_change.cpp

```diff
@@ -0,0 +1,861 @@
+
+        test_rules["roles"][0] = {transform_service_role(general_role)};
+        harness->do_with_new_realm([&](SharedRealm new_realm) {
+            set_up_realm(new_realm, num_total);
+
+            // Add the initial rule and verify the data in realm 1 and 2 (both should just have the employees)
+            update_perms_and_verify(*harness, realm_1, test_rules, {BootstrapMode::Any, params.num_emps, 0, 0});
+            REQUIRE(!wait_for_download(*new_realm));
+            REQUIRE(!wait_for_upload(*new_realm));
+            wait_for_advance(*new_realm);
+            verify_records(new_realm, params.num_emps, 0, 0);
+        });
+        {
+            // Create another user and a new realm config for that user
+            create_user_and_log_in(harness->app());
+            auto config_2 = harness->make_test_file();
+            REQUIRE(config_1.sync_config->user->user_id() != config_2.sync_config->user->user_id());
+            std::atomic<bool> test_started = false;
+
+            // Reopen realm 2 and add a hook callback to check for bootstraps, which should not happen
+            // on this realm
+            config_2.sync_config->on_sync_client_event_hook = [&](std::weak_ptr<SyncSession>,
+                                                                  const SyncClientHookData& data) {
+                using Event = SyncClientHookEvent;
+                if (!test_started.load()) {
+                    return SyncClientHookAction::NoAction; // Not checking yet
+                }
+                // If a download message was received or bootstrap was processed, then fail the test
+                if ((data.event == Event::DownloadMessageReceived &&
+                     data.batch_state != sync::DownloadBatchState::SteadyState) ||
+                    data.event == Event::BootstrapMessageProcessed || data.event == Event::BootstrapProcessed) {
+                    FAIL("Bootstrap occurred on the second realm, which was not expected");
+                }
+                return SyncClientHookAction::NoAction;
+            };
+            auto realm_2 = Realm::get_shared_realm(config_2);
+            set_up_realm(realm_2, params.num_emps);
+
+            test_started = true;
+            // The first rule allows access to all records for user 1
+            AppCreateConfig::ServiceRole user1_role{"user 1 role"};
+            user1_role.apply_when = {{"%%user.id", config_1.sync_config->user->user_id()}};
+            // Add two rules, the first applies to user 1 and the second applies to other users
+            test_rules["roles"] = {transform_service_role(user1_role), transform_service_role(general_role)};
+            update_perms_and_verify(*harness, realm_1, test_rules,
+                                    {BootstrapMode::Any, params.num_emps, params.num_mgrs, params.num_dirs});
+
+            // Realm 2 data should not change (and there shouldn't be any bootstrap messages)
```

Comment:
If there shouldn't be any bootstrap messages, why do we pass BootstrapMode::Any to update_perms_and_verify above? Or am I mis-understanding the meaning of BootstrapMode::Any?

Reply:
`BootstrapMode::Any` means that a bootstrap is expected, but we don't care what type of bootstrap (single message, multi message, etc.) occurs.